json-format.md

```diff
@@ -129,34 +129,82 @@ OPTIONAL not omitted attributes MAY be represeted as a `null` JSON value.
+than as a JSON-formatted string. An implementation MAY fail to serialize the
+event if it is unable to translate the runtime value to a JSON value.
+
+Otherwise, if the `datacontenttype` is NOT JSON-oriented, a JSON serializer
+MUST use the member name `data` to store a string representation of the data
+value, properly encoded according to the `datacontenttype`, in the JSON
+representation. An implementation MAY fail to serialize the event if it is
+unable to represent the runtime value as a properly encoded string.
+
+Out of this follows that presence of the `data` and `data_base64` members is
+mutually exclusive in a JSON serialized CloudEvent.
+
+Furthermore, unlike attributes, for which value types are restricted to strings
+per the [type-system mapping](#22-type-system-mapping), the `data` member
+[JSON value][json-value] is unrestricted, and MAY contain any valid JSON if the
+`datacontenttype` is JSON-oriented. In particular, the the `data` member MAY
+have a value of `null`, representing an explicit `null` payload as distinct from
+the absence of the `data` member.
+
+#### 3.1.2. Payload Deserialization
 
 When a CloudEvents is deserialized from JSON, the presence of the `data_base64`
 member clearly indicates that the value is a Base64 encoded binary data, which
-the serializer MUST decode into a binary runtime data type. When a `data` member
-is present, it is decoded using the default JSON type mapping for the used
-runtime.
-
-Unlike attributes, for which value types are restricted to strings per the
-[type-system mapping](#22-type-system-mapping), the resulting `data` member
-[JSON value][json-value] is unrestricted, and MAY contain any valid JSON.
+the deserializer MUST decode into a binary runtime data type. The deserializer
+MAY further interpret this binary data according to the `datacontenttype`.
+
+When a `data` member is present, the decoding behavior is dependent on the value
+of the `datacontenttype` attribute. If the `datacontenttype` is JSON-oriented
+(that is, its subtype is `json` or has a `+json` format extension), then the
+`data` member MUST be treated directly as a [JSON value][json-value] and decoded
+using the default JSON type mapping for the runtime. Note: if the `data` member
+is a string, a JSON deserializer SHOULD interpret it as a [JSON
+String][json-string] value; it SHOULD NOT further deserialize the string as a
+JSON document.
+
+If the `datacontenttype` is NOT JSON-oriented, then the `data` member SHOULD be
+treated as an encoded content string. An implementation MAY fail to deserialize
+the event if the `data` member is not a string, or if it is unable to interpret
+the `data` with the `datacontenttype`.
+
+If the `datacontenttype` attribute is absent, a JSON deserializer SHOULD proceed
+as if it were set to `application/json`, a JSON-oriented content type.
+Furthermore, when translating an event message with no `datacontenttype`
```

Comment:
For the C# SDK at least, there isn't a "translation" operation - there's "serialize a CloudEvent to bytes" and "deserialize a CloudEvent from bytes". Is the intention that the deserialization operation should set the data content type of the returned CloudEvent to "application/json" if it wasn't set? (Basically, is there *any* way a user should be able to tell the difference between an event deserialized from a JSON format without datacontenttype, and one where datacontenttype was set to "application/json"?)

Reply:
The intent here is that we have a JSON-formatted CE with `datacontenttype` absent, it is deserialized from bytes, and then re-serialized in a different format (say, HTTP binary). The resulting re-serialized CE _must_ have `datacontenttype` set to `application/json` in order to preserve the event semantics. Otherwise, we have lost that information that was implied in the original JSON-formatted serialization.

It is not my intention to say anything about whether a user should be able to tell the difference between the cases you cite. I argue that the two cases are _semantically equivalent_, and I'd also argue that after those two events are re-serialized as HTTP Binary, it would not be possible to distinguish them. But in the intermediate step (e.g. in-memory representation after deserialization from JSON), I don't want to set a requirement one way or the other.

I agree "translating" is probably the wrong term. Maybe better:

> Furthermore, if a JSON-formatted event with no `datacontenttype` is deserialized and then re-serialized in a different format or protocol binding, the `datacontenttype` in the re-serialized event SHOULD be set explicitly to the implied `application/json` content type to preserve the semantics of the event.